scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,444 @@
+
+        let normal_row = &rows[0];
+        let generation_row = &rows[1];
+
+        assert_eq!(normal_row.2.to_std().unwrap(), checkpoint.timestamp);
+        assert_eq!(normal_row.1, checkpoint.generation.timestamp);
+        assert_eq!(generation_row.1, normal_row.1);
+        assert_eq!(normal_row.0, checkpoint.stream_id);
+        assert_eq!(generation_row.0.id, get_default_generation_pk());
+
+        assert_eq!(
+            *cp_saver.last_seen_generation.lock().await,
+            checkpoint.generation
+        );
+    }
+
+    #[tokio::test]
+    async fn test_loading_last_generation_single_producer() {
+        let table_name = unique_name();
+
+        let schema = get_checkpoint_table_schema(&table_name);
+        let (session, ks) = prepare_db(&[schema], 1).await.unwrap();
+
+        let good_generation = GenerationTimestamp {
+            timestamp: chrono::Duration::seconds(420),
+        };
+        let bad_generation1 = GenerationTimestamp {
+            timestamp: chrono::Duration::seconds(200),
+        };
+        let bad_generation2 = GenerationTimestamp {
+            timestamp: chrono::Duration::seconds(600),
+        };
+        let sample_timestamp = Timestamp(chrono::Duration::seconds(42));
+
+        session
+            .query(
+                format!(
+                    "INSERT INTO {} (stream_id, time, generation) VALUES (?, ?, ?)",
+                    table_name
+                ),
+                ("0x00", sample_timestamp, bad_generation1),
+            )
+            .await
+            .unwrap();
+        session
+            .query(
+                format!(
+                    "INSERT INTO {} (stream_id, time, generation) VALUES (?, ?, ?)",
+                    table_name
+                ),
+                ("0xff", sample_timestamp, bad_generation2),
+            )
+            .await
+            .unwrap();
+        session
+            .query(
+                format!(
+                    "INSERT INTO {} (stream_id, time, generation) VALUES (?, ?, ?)",
+                    table_name
+                ),
+                (
+                    get_default_generation_pk(),
+                    sample_timestamp,
+                    good_generation.clone(),
+                ),
+            )
+            .await
+            .unwrap();
+
+        let cp_saver = CDCDefaultCheckpointSaver::new(&session, &ks, Some(&table_name))
+            .await
+            .unwrap();
+
+        let last_generation = cp_saver.load_last_generation().await.unwrap().unwrap();
+        assert_eq!(last_generation, good_generation);
+    }
+
+    // Creates NUMBER_OF_PRODUCERS producers that create unique checkpoints.
+    // Checks whether `load_last_generation` will really return the largest value of seen generations.
+    // Also checks whether number of checkpoints in the table is equal to the number of tasks making checkpoints + one special for generations.
+    #[tokio::test]
+    async fn test_multiple_producers() {
```

Comment:
This test also seems to suffer from the timing issues. I'm not sure that we need such a complicated test; it seems to try and mimic how the library will save checkpoints, and IMO it will be better to write an integration test and use the actual CDCLogReader for that.

In this module, I'd rather see simpler and deterministic tests, each of them testing one characteristic of the checkpoint saver interface. For example:

1. Check that after calling `save_checkpoint` multiple times for one stream `load_last_checkpoint` will return the last checkpoint,
2. Check that after calling `save_checkpoint` once for each stream you will be able to retrieve the same checkpoints for each stream with `load_last_checkpoint`,
3. A similar test to 1) but operating on generations, not on checkpoints,
4. Etc.?

As for the integration test, we can use the end-to-end framework. The test can run a small number of rounds, in each, first write some data and then run the cdc log reader up to a timestamp after the current time. The test cdc log reader should use the checkpoint saver to restore from the previously saved progress, save progress periodically and also do it when the cdc log reader is stopped. At the end, we should make sure that all rows were read and each one of them was read only once (the framework and the test cdc log reader already does that).